### PR TITLE
Optimize the generic string.Join for length 1 collections

### PR DESCRIPTION
**Pull request by jamesqo**
I noticed an optimization made by @bbowyersmyth for `Join` some time ago for single-item lists-- (#1460) but it looks like it wasn't done for the generic version. The one that accepts an `IEnumerable<T>` is a little trickier to optimize since you need to handle the fact that either `currentValue` or its `ToString` could be null, but here it is.

Notes:

- I took extra care to keep the virtual method call order consistent-- everything is called in the order `MoveNext` / `Current` / `ToString`, in case for some bizarre reason one of the latter 2 modify state.

- I also saw what appeared to be some dead code somewhere else in String (`private consts` that are not used anywhere), so I removed that.

CoreFX tests [already exist](https://github.com/dotnet/corefx/blob/master/src/System.Runtime/tests/System/StringTests.cs#L1463) for this overload, including for objects with `ToString` returning null.

cc @jkotas @bbowyersmyth @AlexGhiondea

### CODE REVIEW

File: src/mscorlib/src/System/String.cs [hunk 1]
```diff
@@ -107,31 +107,57 @@ public sealed class String : IComparable, ICloneable, IConvertible, IEnumerable
...
 
-            using(IEnumerator<T> en = values.GetEnumerator()) {
+            using (IEnumerator<T> en = values.GetEnumerator())
+            {
                 if (!en.MoveNext())
-                    return String.Empty;
+                    return string.Empty;
+                
+                // We called MoveNext once, so this will be the first item
+                T firstValue = en.Current;
```
Comment (bbowyersmyth) [inline src/mscorlib/src/System/String.cs]: Nit: Could save a local and declare T currentValue here

File: src/mscorlib/src/System/String.cs [hunk 2]
```diff
@@ -107,31 +107,57 @@ public sealed class String : IComparable, ICloneable, IConvertible, IEnumerable
...
+                }
 
                 StringBuilder result = StringBuilderCache.Acquire();
-                T currentValue = en.Current;
 
-                if (currentValue != null) {
-                    result.Append(currentValue.ToString());
+                // Don't call Append if either firstValue
+                // or its ToString is null (it's a noop)
+                if (firstString != null)
```
Comment (bbowyersmyth) [inline src/mscorlib/src/System/String.cs]: Nit: StringBuilder handles nulls, this condition could removed

**Comment by bbowyersmyth** [comment]
Some minor comments. LGTM

**Comment by jamesqo** [comment]
@bbowyersmyth Thanks for the feedback! I've pushed another commit that fixes the nits.

Just curious, do you still hold the same opinion as in #1460 that the IList-based optimization should not be applied here? If we know the count of the collection beforehand it's possible to call the `Join(string, string[])` overload like this:

```cs
var list = values as IList<T>;
var strings = new string[list.Count];
for (int i = 0; i < strings.Length; i++)
{
    strings[i] = list[i];
}
return Join(separator, strings);
```

which is more efficient since it concats all the strings in one shot (avoiding any intermediary representations).

**Comment by adityamandaleeka** [comment]
The Ubuntu failure is issue #6222.

**Comment by adityamandaleeka** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test

**Comment by jamesqo** [comment]
@jkotas PTAL

@adityamandaleeka Thank you for restarting the CI.

**Comment by adityamandaleeka** [comment]
LGTM. The Ubuntu CI failed again with #6222.

**Pull request merged**